pkg/cvo/availableupdates.go

```diff
@@ -193,71 +217,96 @@ func calculateAvailableUpdatesStatus(ctx context.Context, clusterID string, tran
 		}
 	}
 
-	current, updates, err := cincinnati.NewClient(uuid, transport).GetUpdates(ctx, upstreamURI, arch, channel, currentVersion)
+	current, updates, conditionalUpdates, err := cincinnati.NewClient(uuid, transport).GetUpdates(ctx, upstreamURI, arch, channel, currentVersion)
 	if err != nil {
 		klog.V(2).Infof("Upstream server %s could not return available updates: %v", upstream, err)
 		if updateError, ok := err.(*cincinnati.Error); ok {
-			return cvoCurrent, nil, configv1.ClusterOperatorStatusCondition{
+			return cvoCurrent, nil, nil, configv1.ClusterOperatorStatusCondition{
 				Type: configv1.RetrievedUpdates, Status: configv1.ConditionFalse, Reason: updateError.Reason,
 				Message: fmt.Sprintf("Unable to retrieve available updates: %s", updateError.Message),
 			}
 		}
 		// this should never happen
-		return cvoCurrent, nil, configv1.ClusterOperatorStatusCondition{
+		return cvoCurrent, nil, nil, configv1.ClusterOperatorStatusCondition{
 			Type: configv1.RetrievedUpdates, Status: configv1.ConditionFalse, Reason: "Unknown",
 			Message: fmt.Sprintf("Unable to retrieve available updates: %s", err),
 		}
 	}
 
-	cvoCurrent, err = convertRetrievedUpdateToRelease(current)
-	if err != nil {
-		return cvoCurrent, nil, configv1.ClusterOperatorStatusCondition{
-			Type: configv1.RetrievedUpdates, Status: configv1.ConditionFalse, Reason: "ResponseInvalid",
-			Message: fmt.Sprintf("Invalid recommended update node: %s", err),
-		}
+	return current, updates, conditionalUpdates, configv1.ClusterOperatorStatusCondition{
+		Type:   configv1.RetrievedUpdates,
+		Status: configv1.ConditionTrue,
+
+		LastTransitionTime: metav1.Now(),
 	}
+}
 
-	var cvoUpdates []configv1.Release
-	for _, update := range updates {
-		cvoUpdate, err := convertRetrievedUpdateToRelease(update)
-		if err != nil {
-			return cvoCurrent, nil, configv1.ClusterOperatorStatusCondition{
-				Type: configv1.RetrievedUpdates, Status: configv1.ConditionFalse, Reason: "ResponseInvalid",
-				Message: fmt.Sprintf("Invalid recommended update node: %s", err),
-			}
-		}
-		cvoUpdates = append(cvoUpdates, cvoUpdate)
+func (u *availableUpdates) evaluateConditionalUpdates(ctx context.Context) {
+	if u == nil {
+		return
 	}
 
-	return cvoCurrent, cvoUpdates, configv1.ClusterOperatorStatusCondition{
-		Type:   configv1.RetrievedUpdates,
-		Status: configv1.ConditionTrue,
+	for i, conditionalUpdate := range u.ConditionalUpdates {
+		if errorCondition := evaluateConditionalUpdate(ctx, &conditionalUpdate); errorCondition != nil {
+			meta.SetStatusCondition(&conditionalUpdate.Conditions, *errorCondition)
+			u.removeUpdate(ctx, conditionalUpdate.Release.Image)
+		} else {
+			meta.SetStatusCondition(&conditionalUpdate.Conditions, metav1.Condition{
+				Type:   "Recommended",
+				Status: metav1.ConditionTrue,
+				// FIXME: ObservedGeneration?  That would capture upstream/channel, but not necessarily the currently-reconciling version.
```

Comment:
Not sure what exactly this FIXME indicates or the full impact of not fixing.

Reply:
`metav1.Condition` has an `ObservedGeneration` property, which helps display some idea of freshness by capturing the `metadata.generation` value for the `spec` content that the operator was[](url)